Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -275,4 +290,113 @@ instance (priority := 100) IsPreprimitive.isQuasiPreprimitive [IsPreprimitive M
 
 end Normal
 
+section Finite
+
+namespace IsPreprimitive
+
+variable {M : Type*} [Group M] {α : Type*} [MulAction M α]
+variable {N β : Type*} [Group N] [MulAction N β]
+
+open scoped BigOperators Pointwise
+
+/-- A pretransitive action on a set of prime order is preprimitive -/
+@[to_additive "A pretransitive action on a set of prime order is preprimitive"]
+theorem of_prime_card [hGX : IsPretransitive M α] (hp : Nat.Prime (Nat.card α)) :
+    IsPreprimitive M α := by
+  refine ⟨fun {B} hB ↦ B.subsingleton_or_nontrivial.imp id fun hB' ↦ ?_⟩
+  have : Finite α := (Nat.card_ne_zero.mp hp.ne_zero).2
+  rw [Set.eq_univ_iff_ncard, eq_comm, ← hp.dvd_iff_eq ((Set.one_lt_ncard).mpr hB').ne']
+  exact hB.ncard_dvd_card hB'.nonempty
+
+variable {φ : M → N} {f : α →ₑ[φ] β}
+
+/-- The target of an equivariant map of large image is preprimitive if the source is -/
+@[to_additive "The target of an equivariant map of large image is preprimitive if the source is"]
+theorem of_card_lt [Finite β] [IsPretransitive N β] [IsPreprimitive M α]
+    (hf' : Nat.card β < 2 * (Set.range f).ncard) :
+    IsPreprimitive N β :=  by
+  classical
+  apply IsPreprimitive.mk
+  intro B hB
+  rcases B.eq_empty_or_nonempty with hB' | hB'
+  · left
+    rw [hB']
+    apply Set.subsingleton_empty
+  rw [IsTrivialBlock, or_iff_not_imp_right]
+  intro hB_ne_top
+  -- we need Set.Subsingleton B ↔ Set.ncard B ≤ 1
+  suffices Set.ncard B < 2 by
+    rw [Nat.lt_succ, Set.ncard_le_one_iff] at this
+    exact fun ⦃x⦄ x_1 ⦃y⦄ ↦ this x_1
+  -- We reduce to proving that (Set.range f).ncard ≤ (orbit N B).ncard
+  apply lt_of_mul_lt_mul_right (lt_of_le_of_lt _ hf') (Nat.zero_le _)
+  simp only [← Nat.card_eq_fintype_card, ← hB.ncard_block_mul_ncard_orbit_eq hB']
+  apply Nat.mul_le_mul_left
+  -- We reduce to proving that (Set.range f ∩ g • B).ncard ≤ 1 for every g
+  have hfin := Fintype.ofFinite (Set.range fun g : N ↦ g • B)
+  rw [(hB.isBlockSystem hB').left.ncard_eq_finsum, finsum_eq_sum_of_fintype]
+  apply le_trans (Finset.sum_le_card_nsmul _ _ 1 _)
+  · rw [nsmul_one, Finset.card_univ, ← Set.toFinset_card, ← Set.ncard_eq_toFinset_card',
+      orbit, Nat.cast_id]
+  · · rintro ⟨x, ⟨g, rfl⟩⟩ _
+      suffices Set.Subsingleton (Set.range f ∩ g • B) by
+        rw [Set.ncard_le_one_iff]
+        exact fun {a b} a_1 a_2 ↦ this a_1 a_2
```

Comment:
As above, if we add the lemma `Set.ncard_le_one_iff_subsingleton` this can become just:
```suggestion
      suffices Set.Subsingleton (Set.range f ∩ g • B) by simpa
```